app/cdap/components/NamespaceAdmin/SourceControlManagement/SourceControlManagementModal.tsx

```diff
@@ -0,0 +1,291 @@
+ * the License.
+ */
+
+import {
+  FormControlLabel,
+  FormHelperText,
+  FormLabel,
+  MenuItem,
+  Radio,
+  RadioGroup,
+  Select,
+  TextField,
+} from '@material-ui/core';
+import React, { useEffect, useReducer, useState } from 'react';
+import { FormGroup, Form } from 'reactstrap';
+import styled from 'styled-components';
+import T from 'i18n-react';
+import { createTheme, ThemeProvider } from '@material-ui/core';
+import { ISourceControlManagement, providers, scmAuthType } from './types';
+import { sourceControlManagementFormReducer } from './reducer';
+import PrimaryContainedButton from 'components/shared/Buttons/PrimaryContainedButton';
+import PrimaryOutlinedButton from 'components/shared/Buttons/PrimaryOutlinedButton';
+import StandardModal from 'components/shared/StandardModal';
+
+const PREFIX = 'features.SourceControlManagement.configModal';
+
+const StyledTextField = styled(TextField)`
+  width: 100%;
+`;
+
+const BoldHeader = styled.h2`
+  font-size: 1.4rem !important;
+  font-weight: bold;
+  margin-bottom: 15px;
+  margin-top: 5px;
+`;
+
+const ValidateMessage = styled.div`
+  margin: 5px;
+  ${({ missingFields }) => (missingFields > 0 ? `color: #d15668;` : `color: #4ab63c;`)}
+`;
+
+const StyledForm = styled(Form)`
+  margin-left: 5px;
+`;
+
+const StyledButtonGroup = styled.div`
+  float: right;
+  button {
+    margin-left: 5px;
+  }
+`;
+
+const AsteriskTheme = createTheme({
+  overrides: {
+    MuiFormLabel: {
+      asterisk: {
+        color: '#db3131',
+        '&$error': {
+          color: '#db3131',
+        },
+      },
+    },
+  },
+});
+
+interface ISourceControlManagementModalProps {
+  isOpen: boolean;
+  toggleModal: () => void;
+  onSubmit: (formState: ISourceControlManagement) => void;
+  sourceControlManagementState?: ISourceControlManagement;
+}
+
+interface ISourceControlManagementFormProps {
+  initialSourceControlManagementState?: ISourceControlManagement | null;
+  onSubmit: (formState: ISourceControlManagement) => void;
+  toggleModal: () => void;
+}
+
+const SourceControlManagementForm = ({
+  initialSourceControlManagementState,
+  onSubmit,
+  toggleModal,
+}: ISourceControlManagementFormProps) => {
+  const [formState, dispatch] = useReducer(
+    sourceControlManagementFormReducer,
+    initialSourceControlManagementState
+  );
+
+  const [linkError, setLinkError] = useState(false);
+  const [tokenNameError, setTokenNameError] = useState(false);
+  const [tokenError, setTokenError] = useState(false);
+  const [missingFields, setMissingFields] = useState(0);
+  const [showValidateMessage, setShowValidateMessage] = useState(false);
+
+  useEffect(() => {
+    if (!formState || !formState.auth || !formState.auth.type) {
+      dispatch({
+        type: 'RESET',
+      });
+    }
+  }, []);
+
+  const handleValueChange = (e, type) => {
+    e.preventDefault();
+    dispatch({
+      type,
+      value: e.target.value,
+    });
+  };
+
+  const validateForm = () => {
```

Comment:
This seems a bad way to validate fields...

Reply:
I think all of these things can be derived from state you already have?

Reply:
for instance, in your reducer, if you set a value to be empty, you can set missing fields + 1, etc - i think you should move all your state stuff into that reducer and this will be less messy

Reply:
Agree with @seanfeld Since we have reducer here, we don't need another list of useState hooks

Reply:
added another reducer to address this